storm-core/src/jvm/org/apache/storm/pacemaker/Pacemaker.java

```diff
@@ -0,0 +1,248 @@
+        if (authenticated) {
+            Set<String> pulseIds = new HashSet<>();
+            for (Object key : heartbeats.keySet()) {
+                String k = (String) key;
+                String[] replaceStr = k.replaceFirst(path, "").split("/");
+                String trimmmed = null;
+                for (String str : replaceStr) {
+                    if (!str.equals("")) {
+                        trimmmed = str;
+                        break;
+                    }
+                }
+                if (trimmmed != null && k.indexOf(path) == 0) {
+                    pulseIds.add(trimmmed);
+                }
+            }
+            HBMessageData hbMessageData = HBMessageData.nodes(new HBNodes(new ArrayList(pulseIds)));
+            return new HBMessage(HBServerMessageType.GET_ALL_NODES_FOR_PATH_RESPONSE, hbMessageData);
+        } else {
+            return notAuthorized();
+        }
+    }
+
+    private HBMessage getPulse(String path, boolean authenticated) {
+        if (authenticated) {
+            byte[] details = (byte[]) heartbeats.get(path);
+            LOG.debug("Getting Pulse for path [ {} ]...data [ {} ].", path, details);
+            pacemakerStats.getPulseCount.incrementAndGet();
+            if (details != null) {
+                pacemakerStats.totalSentSize.addAndGet(details.length);
+            }
+            HBPulse hbPulse = new HBPulse();
+            hbPulse.set_id(path);
+            hbPulse.set_details(details);
+            return new HBMessage(HBServerMessageType.GET_PULSE_RESPONSE, HBMessageData.pulse(hbPulse));
+        } else {
+            return notAuthorized();
+        }
+    }
+
+    private HBMessage deletePath(String path) {
+        String prefix = path.endsWith("/") ? path : (path + "/");
+        for (Object key : heartbeats.keySet()) {
+            if (((String) key).indexOf(prefix) == 0)
+                deletePulseId((String) key);
+        }
+        return new HBMessage(HBServerMessageType.DELETE_PATH_RESPONSE, null);
+    }
+
+    private HBMessage deletePulseId(String path) {
+        LOG.debug("Deleting Pulse for id [ {} ].", path);
+        heartbeats.remove(path);
+        return new HBMessage(HBServerMessageType.DELETE_PULSE_ID_RESPONSE, null);
+    }
+
+    private void updateLargestHbSize(int size) {
+        int newValue = size;
+        while (true) {
+            int oldValue = pacemakerStats.largestHeartbeatSize.get();
+            if (newValue > oldValue) {
+                if (!pacemakerStats.largestHeartbeatSize.compareAndSet(oldValue, newValue))
+                    continue;
+            }
+            break;
+        }
+    }
+
+    private void updateAverageHbSize(int size) {
+        int newValue = size;
+        while (true) {
+            int oldValue = pacemakerStats.averageHeartbeatSize.get();
+            int count = pacemakerStats.sendPulseCount.get();
+            newValue = ((count * oldValue) + newValue) / (count + 1);
```

Comment:
This will keep modifying the newValue. I think it should be something like this
` newValue = ((count * oldValue) + size) / (count + 1);`